File: Pinball App 2/Pinball App 2/league/LeaguePreviewParsing.swift
import Foundation

struct LeagueStandingsPreviewPayload {
    var seasonLabel: String = "Season"
    var topRows: [LeagueStandingsPreviewRow] = []
    var aroundRows: [LeagueStandingsPreviewRow] = []
    var currentPlayerStanding: LeagueStandingsPreviewRow?
}

struct LeagueStatsPreviewPayload {
    var rows: [LeagueStatsPreviewRow] = []
    var bankLabel: String = "Most Recent Bank"
    var playerRawName: String = ""
}

func buildLeagueStandingsPreview(standingsCSV: String, selectedPlayer: String?) -> LeagueStandingsPreviewPayload {
    let rows = parseLeagueStandingsRows(standingsCSV)
    guard !rows.isEmpty else {
        return LeagueStandingsPreviewPayload()
    }

    let latestSeason = rows.map(\.season).max() ?? 0
    let seasonLabel = latestSeason > 0 ? "Season \(latestSeason)" : "Season"
    let seasonRows = rows.filter { $0.season == latestSeason }
    guard !seasonRows.isEmpty else {
        return LeagueStandingsPreviewPayload(seasonLabel: seasonLabel)
    }

    let sortedRows: [LeagueParsedStandingRow]
    if seasonRows.allSatisfy({ $0.rank != nil }) {
        sortedRows = seasonRows.sorted { ($0.rank ?? Int.max) < ($1.rank ?? Int.max) }
    } else {
        sortedRows = seasonRows.sorted { $0.total > $1.total }
    }

    let previewRows = sortedRows.enumerated().map { index, row in
        LeagueStandingsPreviewRow(
            rank: row.rank ?? (index + 1),
            rawPlayer: row.player,
            points: row.total
        )
    }

    guard let selectedPlayer, !selectedPlayer.isEmpty else {
        return LeagueStandingsPreviewPayload(
            seasonLabel: seasonLabel,
            topRows: Array(previewRows.prefix(5))
        )
    }

    let normalizedSelected = normalizeLeagueHumanName(selectedPlayer)
    guard let selectedIndex = previewRows.firstIndex(where: {
        normalizeLeagueHumanName($0.rawPlayer) == normalizedSelected
    }) else {
        return LeagueStandingsPreviewPayload(
            seasonLabel: seasonLabel,
            topRows: Array(previewRows.prefix(5))
        )
    }

    return LeagueStandingsPreviewPayload(
        seasonLabel: seasonLabel,
        topRows: Array(previewRows.prefix(5)),
        aroundRows: Array(leagueAroundRowsWindow(rows: previewRows, selectedIndex: selectedIndex)),
        currentPlayerStanding: previewRows[selectedIndex]
    )
}

func buildLeagueStatsPreview(statsCSV: String, preferredPlayer: String?) -> LeagueStatsPreviewPayload {
    let rows = parseLeagueStatsRows(statsCSV)
    guard !rows.isEmpty else {
        return LeagueStatsPreviewPayload()
    }

    let selectedPlayer = resolveLeaguePlayerForStats(preferredPlayer: preferredPlayer, rows: rows)
    let normalizedSelected = normalizeLeagueHumanName(selectedPlayer)
    let selectedRows = rows.filter { normalizeLeagueHumanName($0.player) == normalizedSelected }

    guard !selectedRows.isEmpty else {
        return LeagueStatsPreviewPayload()
    }

    let grouped = Dictionary(grouping: selectedRows, by: { "\($0.season)-\($0.bankNumber)" })
    let mostRecentKey = grouped.keys.max { lhs, rhs in
        guard let lhsRows = grouped[lhs], let rhsRows = grouped[rhs] else { return false }
        return latestLeagueSortValue(lhsRows) < latestLeagueSortValue(rhsRows)
    }

    guard let mostRecentKey,
          let mostRecentRows = grouped[mostRecentKey],
          let sample = mostRecentRows.first else {
        return LeagueStatsPreviewPayload()
    }

    let sortedMostRecentRows = mostRecentRows.sorted { $0.sourceOrder < $1.sourceOrder }
    let rowsForPreview: [LeagueParsedStatsRow] = {
        guard sortedMostRecentRows.count > 5 else { return sortedMostRecentRows }
        let nonZeroScoreRows = sortedMostRecentRows.filter { abs($0.rawScore) > 0.000_001 }
        return nonZeroScoreRows.count >= 5 ? nonZeroScoreRows : sortedMostRecentRows
    }()

    return LeagueStatsPreviewPayload(
        rows: rowsForPreview.prefix(5).enumerated().map { localIndex, row in
            LeagueStatsPreviewRow(
                machine: row.machine,
                score: row.rawScore,
                points: row.points,
                order: localIndex
            )
        },
        bankLabel: "Most Recent • S\(sample.season) B\(sample.bankNumber)",
        playerRawName: sample.player
    )
}

func parseLeagueTargetRows(_ text: String) -> [LeagueTargetPreviewRow] {
    let table = parseCSVRows(text)
    guard let header = table.first else { return [] }

    let headers = header.map(normalizeCSVHeader)
    guard let gameIndex = headers.firstIndex(of: "game"),
          let secondIndex = headers.firstIndex(of: "second_highest_avg"),
          let fourthIndex = headers.firstIndex(of: "fourth_highest_avg"),
          let eighthIndex = headers.firstIndex(of: "eighth_highest_avg") else {
        return []
    }

    return table.dropFirst().compactMap { row in
        guard row.indices.contains(gameIndex),
              row.indices.contains(secondIndex),
              row.indices.contains(fourthIndex),
              row.indices.contains(eighthIndex) else {
            return nil
        }

        let game = row[gameIndex].trimmingCharacters(in: .whitespacesAndNewlines)
        guard !game.isEmpty else { return nil }

        return LeagueTargetPreviewRow(
            game: game,
            secondHighest: Int64(row[secondIndex]) ?? 0,
            fourthHighest: Int64(row[fourthIndex]) ?? 0,
            eighthHighest: Int64(row[eighthIndex]) ?? 0,
            bank: nil,
            order: Int.max
        )
    }
}

func parseLeagueStandingsRows(_ text: String) -> [LeagueParsedStandingRow] {
    let table = parseCSVRows(text)
    guard let header = table.first else { return [] }
    let headers = header.map(normalizeCSVHeader)

    guard let seasonIndex = headers.firstIndex(of: "season"),
          let playerIndex = headers.firstIndex(of: "player"),
          let totalIndex = headers.firstIndex(of: "total") else {
        return []
    }

    let rankIndex = headers.firstIndex(of: "rank")

    return table.dropFirst().compactMap { row in
        guard row.indices.contains(seasonIndex),
              row.indices.contains(playerIndex),
              row.indices.contains(totalIndex) else {
            return nil
        }

        let season = coerceSeasonNumber(row[seasonIndex])
        let player = row[playerIndex].trimmingCharacters(in: .whitespacesAndNewlines)
        let total = Double(row[totalIndex]) ?? 0

        guard season > 0, !player.isEmpty else { return nil }

        let rank: Int?
        if let rankIndex, row.indices.contains(rankIndex) {
            rank = Int(row[rankIndex].trimmingCharacters(in: .whitespacesAndNewlines))
        } else {
            rank = nil
        }

        return LeagueParsedStandingRow(season: season, player: player, total: total, rank: rank)
    }
}

func parseLeagueStatsRows(_ text: String) -> [LeagueParsedStatsRow] {
    let table = parseCSVRows(text)
    guard let header = table.first else { return [] }
    let headers = header.map(normalizeCSVHeader)

    guard let seasonIndex = headers.firstIndex(of: "season"),
          let bankIndex = headers.firstIndex(of: "banknumber"),
          let playerIndex = headers.firstIndex(of: "player"),
          let machineIndex = headers.firstIndex(of: "machine"),
          let scoreIndex = headers.firstIndex(of: "rawscore"),
          let pointsIndex = headers.firstIndex(of: "points") else {
        return []
    }

    let eventDateIndex = headers.firstIndex(of: "eventdate")

    return table.dropFirst().enumerated().compactMap { offset, row in
        guard row.indices.contains(seasonIndex),
              row.indices.contains(bankIndex),
              row.indices.contains(playerIndex),
              row.indices.contains(machineIndex),
              row.indices.contains(scoreIndex),
              row.indices.contains(pointsIndex) else {
            return nil
        }

        let season = coerceSeasonNumber(row[seasonIndex])
        let bankNumber = Int(row[bankIndex].trimmingCharacters(in: .whitespacesAndNewlines)) ?? 0
        let player = row[playerIndex].trimmingCharacters(in: .whitespacesAndNewlines)
        let machine = row[machineIndex].trimmingCharacters(in: .whitespacesAndNewlines)
        let score = Double(row[scoreIndex].replacingOccurrences(of: ",", with: "")) ?? 0
        let points = Double(row[pointsIndex].replacingOccurrences(of: ",", with: "")) ?? 0

        guard season > 0, bankNumber > 0, !player.isEmpty, !machine.isEmpty else { return nil }
        guard score > 0 || points > 0 else { return nil }

        let eventDate: Date?
        if let eventDateIndex, row.indices.contains(eventDateIndex) {
            eventDate = leagueEventDateFormatter.date(from: row[eventDateIndex].trimmingCharacters(in: .whitespacesAndNewlines))
        } else {
            eventDate = nil
        }

        return LeagueParsedStatsRow(
            season: season,
            bankNumber: bankNumber,
            player: player,
            machine: machine,
            rawScore: score,
            points: points,
            eventDate: eventDate,
            sourceOrder: offset
        )
    }
}

func mergeLeagueTargetsWithLibrary(
    targetRows: [LeagueTargetPreviewRow],
    libraryEntries: [LibraryGameLookupEntry]
) -> [LeagueTargetPreviewRow] {
    return targetRows.map { row in
        let bestMatch = LibraryGameLookup.bestMatch(gameName: row.game, entries: libraryEntries)
        guard let bestMatch else { return row }

        return LeagueTargetPreviewRow(
            game: row.game,
            secondHighest: row.secondHighest,
            fourthHighest: row.fourthHighest,
            eighthHighest: row.eighthHighest,
            bank: bestMatch.bank,
            order: bestMatch.order
        )
    }
}

func resolveLeagueNextBank(statsCSV: String?, availableBanks: Set<Int>, preferredPlayer: String?) -> Int? {
    let sortedBanks = availableBanks.sorted()
    guard !sortedBanks.isEmpty else { return nil }
    guard let statsCSV else { return sortedBanks.first }

    let statsRows = parseLeagueStatsRows(statsCSV)
    guard !statsRows.isEmpty else { return sortedBanks.first }

    let scopedRows = scopedLeagueStatsRows(statsRows, preferredPlayer: preferredPlayer)
    guard !scopedRows.isEmpty else { return sortedBanks.first }

    let latestSeason = scopedRows.map(\.season).max() ?? 0
    guard latestSeason > 0 else { return sortedBanks.first }

    let playedBanks = Set(
        scopedRows
            .filter { $0.season == latestSeason && sortedBanks.contains($0.bankNumber) }
            .map(\.bankNumber)
    )

    if let lowestMissing = sortedBanks.first(where: { !playedBanks.contains($0) }) {
        return lowestMissing
    }

    return sortedBanks.first
}

private func resolveLeaguePlayerForStats(preferredPlayer: String?, rows: [LeagueParsedStatsRow]) -> String {
    if let preferredPlayer, !preferredPlayer.isEmpty {
        let normalized = normalizeLeagueHumanName(preferredPlayer)
        if rows.contains(where: { normalizeLeagueHumanName($0.player) == normalized }) {
            return preferredPlayer
        }
    }

    if let latestRow = rows.max(by: { latestLeagueSortValue($0) < latestLeagueSortValue($1) }) {
        return latestRow.player
    }

    return rows[0].player
}

private func latestLeagueSortValue(_ row: LeagueParsedStatsRow) -> Double {
    let dateValue = row.eventDate?.timeIntervalSince1970 ?? 0
    return (dateValue * 1_000_000) + Double(row.season * 100 + row.bankNumber)
}

private func latestLeagueSortValue(_ rows: [LeagueParsedStatsRow]) -> Double {
    rows.map(latestLeagueSortValue).max() ?? 0
}

private func scopedLeagueStatsRows(_ rows: [LeagueParsedStatsRow], preferredPlayer: String?) -> [LeagueParsedStatsRow] {
    guard let preferredPlayer, !preferredPlayer.isEmpty else { return rows }
    let normalizedPreferred = normalizeLeagueHumanName(preferredPlayer)
    let selectedRows = rows.filter { normalizeLeagueHumanName($0.player) == normalizedPreferred }
    return selectedRows.isEmpty ? rows : selectedRows
}

private func normalizeLeagueHumanName(_ raw: String) -> String {
    raw
        .lowercased()
        .components(separatedBy: .whitespacesAndNewlines)
        .filter { !$0.isEmpty }
        .joined(separator: " ")
}

private func leagueAroundRowsWindow(
    rows: [LeagueStandingsPreviewRow],
    selectedIndex: Int,
    windowSize: Int = 5
) -> ArraySlice<LeagueStandingsPreviewRow> {
    guard !rows.isEmpty else { return [] }
    let clampedIndex = max(0, min(selectedIndex, rows.count - 1))
    let edge = windowSize / 2
    let start: Int
    if clampedIndex <= edge {
        start = 0
    } else if clampedIndex >= rows.count - edge - 1 {
        start = max(0, rows.count - windowSize)
    } else {
        start = clampedIndex - edge
    }
    let end = min(rows.count, start + windowSize)
    return rows[start..<end]
}

private let leagueEventDateFormatter: DateFormatter = {
    let formatter = DateFormatter()
    formatter.calendar = Calendar(identifier: .gregorian)
    formatter.locale = Locale(identifier: "en_US_POSIX")
    formatter.dateFormat = "yyyy-MM-dd"
    return formatter
}()

struct LeagueParsedStandingRow {
    let season: Int
    let player: String
    let total: Double
    let rank: Int?
}

struct LeagueParsedStatsRow {
    let season: Int
    let bankNumber: Int
    let player: String
    let machine: String
    let rawScore: Double
    let points: Double
    let eventDate: Date?
    let sourceOrder: Int
}

File: Pinball App Android/app/src/main/java/com/pillyliu/pinprofandroid/league/LeaguePreviewParsing.kt
package com.pillyliu.pinprofandroid.league

import com.pillyliu.pinprofandroid.data.parseCsv
import com.pillyliu.pinprofandroid.library.LibraryGameLookup
import com.pillyliu.pinprofandroid.library.LibraryGameLookupEntry
import java.time.LocalDate
import java.time.format.DateTimeFormatter
import java.util.Locale
import kotlin.math.abs

internal fun buildStatsPreview(rows: List<StatsCsvRow>, preferredPlayer: String?): StatsPreviewPayload {
    if (rows.isEmpty()) return StatsPreviewPayload(emptyList(), "Most Recent Bank", "")

    val selectedPlayer = resolveLeaguePlayerForStats(rows, preferredPlayer)
    val selected = rows.filter { matchesLeaguePreferredPlayer(it.player, selectedPlayer) }
    if (selected.isEmpty()) return StatsPreviewPayload(emptyList(), "Most Recent Bank", "")

    val grouped = selected.groupBy { "${it.season}-${it.bank}" }
    val recentKey = grouped.keys.maxByOrNull { key ->
        grouped[key]?.maxOfOrNull(::latestLeagueSortValue) ?: 0L
    } ?: return StatsPreviewPayload(emptyList(), "Most Recent Bank", "")
    val sample = grouped[recentKey]?.firstOrNull() ?: return StatsPreviewPayload(emptyList(), "Most Recent Bank", "")

    val sortedMostRecentRows = grouped[recentKey].orEmpty().sortedBy { it.sourceOrder }
    val rowsForPreview = if (sortedMostRecentRows.size > 5) {
        val nonZeroScoreRows = sortedMostRecentRows.filter { abs(it.score) > 0.000001 }
        if (nonZeroScoreRows.size >= 5) nonZeroScoreRows else sortedMostRecentRows
    } else {
        sortedMostRecentRows
    }

    return StatsPreviewPayload(
        rows = rowsForPreview.take(5).map {
            StatsPreviewRow(machine = it.machine, score = it.score, points = it.points)
        },
        bankLabel = "Most Recent • S${sample.season} B${sample.bank}",
        playerRawName = sample.player,
    )
}

internal fun buildStandingsPreview(rows: List<StandingCsvRow>, selectedPlayer: String?): StandingsPreviewPayload {
    if (rows.isEmpty()) return StandingsPreviewPayload("Season", emptyList(), emptyList())
    val latestSeason = rows.maxOfOrNull { it.season } ?: return StandingsPreviewPayload("Season", emptyList(), emptyList())
    val seasonRows = rows.filter { it.season == latestSeason }
    if (seasonRows.isEmpty()) return StandingsPreviewPayload("Season $latestSeason", emptyList(), emptyList())

    val sorted = if (seasonRows.all { it.rank != null }) {
        seasonRows.sortedBy { it.rank ?: Int.MAX_VALUE }
    } else {
        seasonRows.sortedByDescending { it.total }
    }
    val previewRows = sorted.mapIndexed { index, row ->
        StandingsPreviewRow(rank = row.rank ?: (index + 1), rawPlayer = row.player, points = row.total)
    }
    val topRows = previewRows.take(5)

    if (!selectedPlayer.isNullOrBlank()) {
        val selectedIndex = previewRows.indexOfFirst { matchesLeaguePreferredPlayer(it.rawPlayer, selectedPlayer) }
        if (selectedIndex >= 0) {
            return StandingsPreviewPayload("Season $latestSeason", topRows, leagueAroundRowsWindow(previewRows, selectedIndex))
        }
    }

    return StandingsPreviewPayload("Season $latestSeason", topRows, emptyList())
}

internal fun parseStatsRows(text: String): List<StatsCsvRow> {
    val table = parseCsv(text)
    if (table.isEmpty()) return emptyList()
    val header = table.first().map(::normalizeLeagueHeader)
    val seasonIndex = header.indexOf("season")
    val bankIndex = header.indexOf("banknumber")
    val playerIndex = header.indexOf("player")
    val machineIndex = header.indexOf("machine")
    val scoreIndex = header.indexOf("rawscore")
    val pointsIndex = header.indexOf("points")
    val eventDateIndex = header.indexOf("eventdate")

    if (listOf(seasonIndex, bankIndex, playerIndex, machineIndex, scoreIndex, pointsIndex).any { it < 0 }) {
        return emptyList()
    }

    return table.drop(1).mapIndexedNotNull { idx, row ->
        if (listOf(seasonIndex, bankIndex, playerIndex, machineIndex, scoreIndex, pointsIndex).any { it !in row.indices }) {
            return@mapIndexedNotNull null
        }
        val season = coerceLeagueSeason(row[seasonIndex])
        val bank = row[bankIndex].trim().toIntOrNull() ?: 0
        val player = row[playerIndex].trim()
        val machine = row[machineIndex].trim()
        val score = row[scoreIndex].trim().replace(",", "").toDoubleOrNull() ?: 0.0
        val points = row[pointsIndex].trim().replace(",", "").toDoubleOrNull() ?: 0.0
        val eventDate = if (eventDateIndex in row.indices) {
            runCatching { LocalDate.parse(row[eventDateIndex].trim(), DateTimeFormatter.ISO_LOCAL_DATE) }.getOrNull()
        } else {
            null
        }

        if (season <= 0 || bank <= 0 || player.isBlank() || machine.isBlank()) return@mapIndexedNotNull null
        if (score <= 0.0 && points <= 0.0) return@mapIndexedNotNull null

        StatsCsvRow(season, bank, player, machine, score, points, eventDate, idx)
    }
}

internal fun parseStandingsRows(text: String): List<StandingCsvRow> {
    val table = parseCsv(text)
    if (table.isEmpty()) return emptyList()
    val header = table.first().map(::normalizeLeagueHeader)
    val seasonIndex = header.indexOf("season")
    val playerIndex = header.indexOf("player")
    val totalIndex = header.indexOf("total")
    val rankIndex = header.indexOf("rank")
    if (listOf(seasonIndex, playerIndex, totalIndex).any { it < 0 }) return emptyList()

    return table.drop(1).mapNotNull { row ->
        if (listOf(seasonIndex, playerIndex, totalIndex).any { it !in row.indices }) return@mapNotNull null
        val season = coerceLeagueSeason(row[seasonIndex])
        val player = row[playerIndex].trim()
        val total = row[totalIndex].trim().replace(",", "").toDoubleOrNull() ?: 0.0
        val rank = if (rankIndex in row.indices) row[rankIndex].trim().toIntOrNull() else null
        if (season <= 0 || player.isBlank()) return@mapNotNull null
        StandingCsvRow(season, player, total, rank)
    }
}

internal fun parseTargetRows(text: String): List<TargetPreviewRow> {
    val table = parseCsv(text)
    if (table.isEmpty()) return emptyList()
    val header = table.first().map(::normalizeLeagueHeader)
    val gameIndex = header.indexOf("game")
    val secondIndex = header.indexOf("second_highest_avg")
    val fourthIndex = header.indexOf("fourth_highest_avg")
    val eighthIndex = header.indexOf("eighth_highest_avg")
    if (listOf(gameIndex, secondIndex, fourthIndex, eighthIndex).any { it < 0 }) return emptyList()

    return table.drop(1).mapNotNull { row ->
        if (listOf(gameIndex, secondIndex, fourthIndex, eighthIndex).any { it !in row.indices }) return@mapNotNull null
        val game = row[gameIndex].trim()
        if (game.isBlank()) return@mapNotNull null
        TargetPreviewRow(
            game = game,
            second = row[secondIndex].trim().toLongOrNull() ?: 0L,
            fourth = row[fourthIndex].trim().toLongOrNull() ?: 0L,
            eighth = row[eighthIndex].trim().toLongOrNull() ?: 0L,
            bank = null,
            order = Int.MAX_VALUE,
        )
    }
}

internal fun mergeTargetsWithLibrary(
    targetRows: List<TargetPreviewRow>,
    libraryEntries: List<LibraryGameLookupEntry>,
): List<TargetPreviewRow> {
    return targetRows.map { row ->
        val match = LibraryGameLookup.bestMatch(row.game, libraryEntries)
        if (match == null) row else row.copy(bank = match.bank, order = match.order)
    }
}

internal fun resolveNextBank(statsRows: List<StatsCsvRow>, availableBanks: Set<Int>, preferredPlayer: String?): Int? {
    val sorted = availableBanks.sorted()
    if (sorted.isEmpty()) return null
    if (statsRows.isEmpty()) return sorted.first()

    val scopedRows = scopedLeagueStatsRows(statsRows, preferredPlayer)
    if (scopedRows.isEmpty()) return sorted.first()

    val latestSeason = scopedRows.maxOfOrNull { it.season } ?: return sorted.first()
    val played = scopedRows
        .filter { it.season == latestSeason && it.bank in sorted }
        .map { it.bank }
        .toSet()

    return sorted.firstOrNull { it !in played } ?: sorted.first()
}

internal fun matchesLeaguePreferredPlayer(candidate: String, preferred: String): Boolean {
    return normalizeLeagueHumanName(candidate) == normalizeLeagueHumanName(preferred)
}

private fun resolveLeaguePlayerForStats(rows: List<StatsCsvRow>, preferredPlayer: String?): String {
    if (!preferredPlayer.isNullOrBlank() && rows.any { matchesLeaguePreferredPlayer(it.player, preferredPlayer) }) {
        return preferredPlayer
    }
    return rows.maxByOrNull(::latestLeagueSortValue)?.player ?: rows.first().player
}

private fun latestLeagueSortValue(row: StatsCsvRow): Long {
    val datePart = row.eventDate?.toEpochDay() ?: 0L
    return (datePart * 1_000_000L) + (row.season * 100L + row.bank)
}

private fun scopedLeagueStatsRows(rows: List<StatsCsvRow>, preferredPlayer: String?): List<StatsCsvRow> {
    if (preferredPlayer.isNullOrBlank()) return rows
    val selectedRows = rows.filter { matchesLeaguePreferredPlayer(it.player, preferredPlayer) }
    return if (selectedRows.isEmpty()) rows else selectedRows
}

private fun normalizeLeagueHeader(value: String): String {
    return value.replace("\uFEFF", "").replace("\u0000", "").trim().lowercase(Locale.US)
}

private fun coerceLeagueSeason(raw: String): Int {
    val trimmed = raw.trim()
    val digits = trimmed.filter { it.isDigit() }
    return digits.toIntOrNull() ?: trimmed.toIntOrNull() ?: 0
}

private fun normalizeLeagueHumanName(raw: String): String {
    val normalized = raw.lowercase(Locale.US)
        .map { ch -> if (ch.isLetterOrDigit() || ch.isWhitespace()) ch else ' ' }
        .joinToString(separator = "")
    return normalized.trim().split(Regex("\\s+")).filter { it.isNotBlank() }.joinToString(" ")
}

private fun leagueAroundRowsWindow(
    rows: List<StandingsPreviewRow>,
    selectedIndex: Int,
    windowSize: Int = 5,
): List<StandingsPreviewRow> {
    if (rows.isEmpty()) return emptyList()
    val clampedIndex = selectedIndex.coerceIn(0, rows.lastIndex)
    val edge = windowSize / 2
    val start = when {
        clampedIndex <= edge -> 0
        clampedIndex >= rows.size - edge - 1 -> maxOf(0, rows.size - windowSize)
        else -> clampedIndex - edge
    }
    val end = minOf(rows.size, start + windowSize)
    return rows.subList(start, end)
}
